src/components/vapp.ts

```diff
@@ -0,0 +1,263 @@
+  vms: VmData[];
+}
+
+/**
+ * Virtual Application Visual Component.
+ */
+export class VappComponent extends paper.Group {
+
+  readonly label: EntityLabelComponent;
+  readonly background: paper.Path.Rectangle;
+  readonly _margin: MarginComponent;
+  readonly vms: VmAndVnicListComponent;
+  readonly vappNetworks: VappNetworkListComponent;
+  // position for the division between any labels and vm/vnic list
+  readonly divisionPositionY: number;
+  private scrollbar: ScrollbarComponent;
+  // content needs scrollbar
+  private _isScrollable: boolean = false;
+
+  /**
+   * Creates a new VappComponent instance.
+   *
+   * @param _vapp the vapp data
+   * @param _point the location that the vm should be rendered at
+   */
+  constructor(private _vapp: VappData,
+              private _point: paper.Point = new paper.Point(0, 0)) {
+    super();
+    this.applyMatrix = false;
+    this.pivot = new paper.Point(0, 0);
+    this.position = _point;
+
+    let backgroundOffsetX = 0;
+    let backgroundOffsetY = 0;
+    const vappNetworkCount = this._vapp.vapp_networks.length;
+    const vmCount = this._vapp.vms.length;
+
+    // vapp label
+    // x position based on if there are any vApp networks (and how many) or vms
+    const labelPositionX = vappNetworkCount || vmCount
+      ? Math.max(1, vappNetworkCount) * VAPP_NETWORK_RIGHT_MARGIN + VAPP_PADDING
+      : VAPP_PADDING;
+    this.label = new EntityLabelComponent(
+      this._vapp.name,
+      LABEL_ICON_COLOR,
+      new paper.Point(labelPositionX, VAPP_PADDING),
+      'bold'
+    );
+    this.addChild(this.label);
+
+    // start vapp network paths - add a point to each network to have x position data for matching vnics
+    // and edge labels to have y position data for vapp header (which includes the vapp label and edge labels)
+    if (vappNetworkCount) {
+      this.vappNetworks = new VappNetworkListComponent(
+        this._vapp.vapp_networks,
+        new paper.Point(VAPP_PADDING + CONNECTOR_RADIUS, 0));
+      this.addChild(this.vappNetworks);
+    }
+
+    // calculate division position between the vapp header (which includes the vapp label and edge labels) and vm
+    // and vnic list
+    const headerBase = this.globalToLocal(this.bounds.bottomLeft).y;
+    // headerBase will equal vapp label bottom if there are no edge labels. margin based on which label is positionally
+    // the lowest.
+    const vappLabelIsBottomMost = headerBase === this.label.bounds.bottom;
+    const headerBaseMargin = vappLabelIsBottomMost ? VAPP_LABEL_BOTTOM_MARGIN : EDGE_LABEL_BOTTOM_MARGIN;
+    this.divisionPositionY = headerBase + headerBaseMargin;
+
+    // maximum bottom position for vApp background based on the canvas view size
+    const maxBottomPosition = this.globalToLocal(paper.view.bounds.bottomLeft).y - VIEW_BOTTOM_PADDING;
+
+    // vms and vnics list
+    this.vms = new VmAndVnicListComponent(
+      this._vapp.vms,
+      this.vappNetworks && this.vappNetworks.networkPositionsByName,
```

Comment:
passing 'networkPositionsByName' derived from vappNetworks component here. is there a better way or naming convention to use? more documentation? `// networkPositionsByName have x position data for matching vnics`

Reply:
since it's clear, i'll leave it as is. there's a possibility that positions might need to be converted (local, global, parent, child) later on, so i'll keep full position data